### PR TITLE
TH2-2061. Implemented retry mechanism for both storing and getting data

### PR DESCRIPTION
**Pull request by Vladimir-Panarin-Exactpro**

### CODE REVIEW

**Comment by Vladimir-Panarin-Exactpro** [comment]
`maxParallelQueries` is not used now and is marked as deprecated
New settings:
- `maxSyncRetries` - number of retries to make in synchronous (blocking) requests. 3 by default
- `maxAsyncRetries` - number of retries to make in asynchronous (non-blocking) requests. -1 by default, i.e. infinite "retry till success"
- `retryDelay` - delay between retries in milliseconds. 100 by default;
- `composingService` - `ExecutorService` used to compose a sequence of the Cassandra query executions. If not defined, `Executors.newFixedThreadPool(5)` will be used and shut down on `CradleStorage` disposal.

If a request (e.g. `storeMessageBatchAsync()` method call) implies multiple Cassandra queries, they are chained and not perform in parallel. I.e. Cradle API user can check the number of incomplete requests to throttle further requests.

Sync requests are blocking current thread till result is available.
Async requests are queued and return non-blocking. Queue with requests is processed via separate thread, sending query to Cassandra asynchronously and assigning callback to perform retry. To retry the request, it is put to the same queue.

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/retry/AsyncExecutor.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	private final AsyncRequestProcessor processor;
+	
+	public AsyncExecutor(int maxQueueSize, ExecutorService composingService, int defaultRetries, int delay)
+	{
+		this.requests = new LinkedBlockingQueue<RequestInfo<?>>(maxQueueSize);
+		this.execService = Executors.newSingleThreadExecutor();
+		this.defaultRetries = defaultRetries;
+		this.processor = new AsyncRequestProcessor(requests, composingService, delay);
+		
+		execService.submit(processor);
```
Comment (OptimumOpium) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/retry/AsyncExecutor.java:48]: How about using ScheduledExecutorService and submit the processor task with a fixed delay between executions? I will allow us to get rid of the "infinite" loop in the `run` method in AsyncRequestProcessor

Comment (Vladimir-Panarin-Exactpro) [reply cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/retry/AsyncExecutor.java:48]: I'm not sure, but it looks like performance of writes will suffer from this if writes occur from time to time.

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/retry/AsyncExecutor.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	}
+	
+	public <T> CompletableFuture<T> submit(String requestInfo, int maxRetries, Supplier<CompletableFuture<T>> supplier)
+	{
+		CompletableFuture<T> result = new CompletableFuture<T>();
+		if (!requests.offer(new RequestInfo<T>(requestInfo, maxRetries, supplier, result)))
+			result.completeExceptionally(new TooManyRequestsException("Could not submit new request '"+requestInfo+"'"));
```
[the remark this answers is not in the document]
Comment (Vladimir-Panarin-Exactpro) [reply cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/retry/AsyncExecutor.java:55]: Actually, I prefer the behavior you described.

**Review by MrJonniK — APPROVED** [review]

**Pull request closed without merging**